gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/publisher/CopyDataPublisher.java

```diff
@@ -225,8 +226,13 @@ private void publishFileSet(CopyEntity.DatasetAndPartition datasetAndPartition,
     Preconditions.checkArgument(!datasetWorkUnitStates.isEmpty(),
         "publishFileSet received an empty collection work units. This is an error in code.");
 
-    CopyableDatasetMetadata metadata = CopyableDatasetMetadata
-        .deserialize(datasetWorkUnitStates.iterator().next().getProp(CopySource.SERIALIZED_COPYABLE_DATASET));
+    CopyableDatasetMetadata metadata = CopyableDatasetMetadata.deserialize(
+        datasetWorkUnitStates.iterator().next().getProp(CopySource.SERIALIZED_COPYABLE_DATASET));
+
+    // Get the dataset specific sharded directory if configured
+    if (state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR)) {
+      this.writerOutputDir = new Path(datasetWorkUnitStates.iterator().next().getProp(ConfigurationKeys.WRITER_OUTPUT_DIR));
+    }
     Path datasetWriterOutputPath = new Path(this.writerOutputDir, datasetAndPartition.identifier());
```

Comment:
1. Can we store `datasetWorkUnitStates.iterator().next()` in a temp variable like `sampledWUS` to avoid calling twice? 
2. `writerOutputDir` in this case should be a local variable since it will be specific to a workunitstate. Maybe converting the whole thing including line 236 into a ternary operator like `Path datasetWriterOutputPath = state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR) ? xxx : this.writerOutputDir` ? just leave the `writerOutputDir` as null otherwise might be safer: we can risk fail the job but not writing something that are not supposed to be in a dataset folder